source/common/common/token_bucket.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <chrono>
+
+namespace Envoy {
+
+/**
+ * This class implements the token bucket algorithm (not thread-safe).
+ *
+ * https://en.wikipedia.org/wiki/Token_bucket
+ */
+class TokenBucket {
+
+  constexpr static double nano = double(std::micro::num) / double(std::micro::den);
```

Comment:
I think you can do all this stuff with `std::chrono` conversions, but not sure, maybe take a look?